Docs: Update create block type how to guide for block.json (#37674)
WordPress/gutenberg: docs/how-to-guides/block-tutorial/writing-your-first-block-type.md

## Overview

There are two main types of blocks: static and dynamic, this guide focuses on static blocks. A static block is used to insert HTML content into the post and save it with the post. A dynamic block builds the content on the fly when rendered on the front end, see the [dynamic blocks guide](/docs/how-to-guides/block-tutorial/creating-dynamic-blocks.md).

This guide focuses on just the block, see the [Create a Block tutorial](/docs/getting-started/create-block/README.md) for a complete setup.

## Before you start

Static blocks are implemented in JavaScript, so a basic level of JavaScript is helpful, see the [Getting Started with JavaScript](/docs/how-to-guides/javascript/README.md) for a refresher.

Blocks are added to WordPress using plugins, so you will need:

-   WordPress development environment, see [setup guide](/docs/getting-started/devenv/README.md)
-   JavaScript build tools (node/npm) if using JSX example

## Step-by-step guide

### Step 1: Configure block.json

The functions of a static block is defined in JavaScript, however the settings and other metadata should be defined in a block.json file.

Here are the basic settings:

> -   `title`: Block title shown in inserter

Reviewer: ```suggestion
-   `apiVersion`: Block API version
-   `title`: Block title shown in inserter
```

Create a basic `block.json` file there:

{% codetabs %}
{% JSX %}

```json
{
	"apiVersion": 2,
	"name": "gutenberg-examples/example-01-basic-esnext",
	"title": "Example: Basic (ESNext)",
	"icon": "universal-access-alt",
	"category": "layout",
	"editorScript": "file:./build/index.js"
}
```

{% Plain %}

```json
{
	"apiVersion": 2,
	"title": "Example: Basic",
	"name": "gutenberg-examples/example-01-basic",
	"category": "layout",
	"icon": "universal-access-alt",
	"editorScript": "file:./block.js"
}
```

{% end %}

### Step 2: Register block in plugin

With the `block.json` in place, the registration for the block is a single function call in PHP, this will setup the block and JavaScript file specified in the `editorScript` property to load in the editor.

Create a full plugin file, `index.php` like the following, the same PHP code works for both JSX and Plain code.

```php
<?php
/**
 * Plugin Name: Gutenberg examples 01
 */
function gutenberg_examples_01_register_block() {
	register_block_type( __DIR__ );
}
add_action( 'init', 'gutenberg_examples_01_register_block' );
```

### Step 3: Block edit and save functions

> The `editorScript` that loads in the editor calls a register function with a matching `name` specified in `block.json` and defines two important functions for the block, the `edit` and `save` functions.

Reviewer: This is a little unclear to me. I wonder if we only need to explain at a high level?
```suggestion
The `editorScript` entry is enqueued automatically in the block editor. This file contains the JavaScript portion of the block registration and defines two important functions for the block, the `edit` and `save` functions.
```

# Create a basic block

This guide takes you through creating a basic block to display a message in a post. This message will be fixed, we won't allow the user to edit the message, the goal of the guide is to show how to register and load a block.

## Overview

> There are two main types of blocks: static and dynamic, this guide focusing on static blocks. A static block is used to insert HTML content into the post and saved with the post. A dynamic block builds the content on the fly when rendered on the front end, see the [dynamic blocks guide](/docs/how-to-guides/block-tutorial/creating-dynamic-blocks.md).

Reviewer 2: I had some trouble processing this section in the first go, would it help to apply some small tweaks like:
```suggestion
There are two main types of blocks: static and dynamic, this guide focuses on static blocks. A static block is used to insert HTML content into the post and save it with the post. A dynamic block builds the content on the fly when rendered on the front end, see the [dynamic blocks guide](/docs/how-to-guides/block-tutorial/creating-dynamic-blocks.md).
```

This guide focuses on just the block, see the [Create a Block tutorial](/docs/getting-started/create-block/README.md) for a complete setup.

## Before you start

Static blocks are implemented in JavaScript, so a basic level of JavaScript is helpful, see the [Getting Started with JavaScript](/docs/how-to-guides/javascript/README.md) for a refresher.

Blocks are added to WordPress using plugins, so you will need:

-   WordPress development environment, see [setup guide](/docs/getting-started/devenv/README.md)
-   JavaScript build tools (node/npm) if using JSX example

## Step-by-step guide

### Step 1: Configure block.json

The functions of a static block is defined in JavaScript, however the settings and other metadata should be defined in a block.json file.

Here are the basic settings:

-   `title`: Block title shown in inserter
-   `name`: Unique name defines your block

> -   `category`: Category in inserter (text, media, design, widgets, theme embed)

Reviewer 2: Typo:
```suggestion
-   `category`: Category in inserter (text, media, design, widgets, theme, embed)
```

-   `icon`: Dashicon icon displayed for block
-   `editorScript`: JavaScript file to load for block

> The block.json file should be added to your plugin. To start a new plugin, create a directory in `/wp-content/plugins/` in your WordPress.

Reviewer 2: Code hightlights:
```suggestion
The `block.json` file should be added to your plugin. To start a new plugin, create a directory in `/wp-content/plugins/` in your WordPress.
```

> Create a basic block.json file there:

Reviewer 2: Code hightlights:
```suggestion
Create a basic `block.json` file there:
```

{% codetabs %}
{% JSX %}

```json
{
	"apiVersion": 2,
	"name": "gutenberg-examples/example-01-basic-esnext",
	"title": "Example: Basic (ESNext)",
	"icon": "universal-access-alt",
	"category": "layout",
	"example": {},
	"editorScript": "file:./build/index.js"
}
```

{% Plain %}

```json
{
	"apiVersion": 2,
	"title": "Example: Basic",
	"name": "gutenberg-examples/example-01-basic",
	"category": "layout",

> 	"icon": "universal-access-alt",

Reviewer 2: Any particular reason why `example` is missing here?

Author: Good spot, I wanted to remove it from both. The goal of the guide is a minimal block with as little extra as needed, to introduce to how block loads and the major `edit` and `save` functions.

	"editorScript": "file:./block.js"
}
```

{% end %}

### Step 2: Register block in plugin

> With the block.json in place, the registration for the block is a single function call in PHP, this will setup the block and JavaScript file specified in the `editorScript` property to load in the editor.

Reviewer 2: Code highlight for conistency:
```suggestion
With the `block.json` in place, the registration for the block is a single function call in PHP, this will setup the block and JavaScript file specified in the `editorScript` property to load in the editor.
```

```php
	register_block_type( __DIR__ );
```

Create a full plugin file, `index.php` like the following:

```php

> <?php

Reviewer 2: Is this plugin metadata comment valid?

This is how it looks like in the examples:

https://github.com/WordPress/gutenberg-examples/blob/86245536fa2bd057282543ce94a7ea9a08974b2f/01-basic-esnext/index.php#L3-L11

https://github.com/WordPress/gutenberg-examples/blob/86245536fa2bd057282543ce94a7ea9a08974b2f/01-basic/index.php#L3-L11

For `block.json` there are two version uses. We should either use a unified version for `index.php` and `block.json` or provide "codetabs" with JSX and Plain here as well.

/*
Plugin Name: Gutenberg examples 01
*/
function gutenberg_examples_01_register_block() {
	register_block_type( __DIR__ );
}
add_action( 'init', 'gutenberg_examples_01_register_block' );
```

### Step 3: Block edit and save functions

> The `editorScript` that loads in the editor calls a register function with a matching `name` specified in block.json and defines two important functions for the block, the `edit` and `save` functions.

Reviewer 2: Code hightlight:
```suggestion
The `editorScript` that loads in the editor calls a register function with a matching `name` specified in `block.json` and defines two important functions for the block, the `edit` and `save` functions.
```

The `edit` function is a component that is shown in the editor when the block is inserted.

The `save` function is a component that defines the final markup returned by the block and saved in `post_content`.

{% codetabs %}
{% JSX %}

Add the following in `src/index.js`

```jsx
/**
 * WordPress dependencies
 */
import { registerBlockType } from '@wordpress/blocks';

// Register the block
registerBlockType( 'gutenberg-examples/example-01-basic-esnext', {
	edit: function () {
		return <p> Hello world (from the editor)</p>;
	},
	save: function () {
		return <p> Hola mundo (from the frontend) </p>;
	},
} );
```

{% Plain %}

Add the following to `block.js`

```js
( function ( blocks, element ) {
	var el = element.createElement;

	blocks.registerBlockType( 'gutenberg-examples/example-01-basic', {
		edit: function () {
			return el( 'p', {}, 'Hello World (from the editor).' );
		},
		save: function () {

> 			var blockProps = useBlockProps.save( { style: blockStyle } );

Reviewer 2: Should it be removed?
```suggestion
```

			return el( 'p', {}, 'Hola mundo (from the frontend).' );
		},
	} );
} )( window.wp.blocks, window.wp.element );
```

{% end %}

NOTE: If using the JSX version, you need to run `npm run build` and it will create the JavaScript file that is loaded in the editor at `build/index.js`

### Step 4: Confirm

Open your editor and try adding your new block. It will show in the inserter using the `title`.
When inserted you will see the `Hello World (from the editor)` message.

> When you save the post and view it published, you will see the `Hello World (from the frontend)` message.

Reviewer 2: It prints a different message on the frontend 😄: 
```suggestion
When you save the post and view it published, you will see the `Hola mundo (from the frontend)` message.
```
